### PR TITLE
feat: inital commit for #152 transfer claim feature

### PR DESCRIPTION
**Pull request by wm798222**
There is an error in the claim transfer test (with existing trust relationships, and specified tokens). Not sure what's wrong…

 1) transferRouter
       claim transfer with existing trust relationship, and specified tokens:
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/wm/Desktop/CS/greenstand-folder/fork/treetracker-wallet-api/server/routes/transferRouter.spec.js)

catch error: Error: Can not rollback transaction before start it!

### CODE REVIEW

**Comment by dadiorchen** [comment]
@wm798222 you are testing the transferTroute.spec.js when you get the error above? If this is true, then you need to check the mock stub, in this test, it shouldn't invoke really DB stuff, so you must mock out all those things. This might be caused by the Session object I think.

**Comment by wm798222** [comment]
> @wm798222 you are testing the transferTroute.spec.js when you get the error above? If this is true, then you need to check the mock stub, in this test, it shouldn't invoke really DB stuff, so you must mock out all those things. This might be caused by the Session object I think.

Hi Dadior, you are right that it is caused by DB issues. In the transfer post route, there is `await session.beginTransaction();`, which will call `this.thx = await knex.transaction();` and it hangs here. I think knex.transaction is calling the DB? Do you think there is a way to get around it given that `await session.beginTransaction();` is called in the post route?

**Comment by dadiorchen** [comment]
> > @wm798222 you are testing the transferTroute.spec.js when you get the error above? If this is true, then you need to check the mock stub, in this test, it shouldn't invoke really DB stuff, so you must mock out all those things. This might be caused by the Session object I think.
> 
> Hi Dadior, you are right that it is caused by DB issues. In the transfer post route, there is `await session.beginTransaction();`, which will call `this.thx = await knex.transaction();` and it hangs here. I think knex.transaction is calling the DB? Do you think there is a way to get around it given that `await session.beginTransaction();` is called in the post route?

@wm798222 definitely possible to work around this, jest can mock out the whole class like Session. Please follow this guide to mock the Session class: https://jestjs.io/docs/en/es6-class-mocks#automatic-mock
There are different way to mock class, in this case, the auto mock is okay I think, because you don't need to mock any return/resolve right?

File: server/models/Wallet.js [hunk 1]
```diff
@@ -487,6 +487,8 @@ class Wallet{
         parameters: {
           tokens: tokensUUID,
         },
+        // TODO: add boolean for claim in transferRepository
+        claim: claimBoolean,
```
Comment (deepwinter) [inline server/models/Wallet.js:491]: This is the unit of logic you really want to test

**Comment by wm798222** [comment]
End to end test performed:

Bundle: Claimed tokens cannot be transferred
1. Wallet does not have enough tokens that are NOT claimed. -> throw an error stating there is a token that is claimed.
2. Wallet has enough tokens that are NOT claimed, claim set to true -> tokens transferred successfully and cannot participate in further transfers
3. Wallet has enough tokens that are NOT claimed, claim set to false -> tokens transferred successfully

Explicit token
1. Claimed token, claim set to false-> can't be transferred
2. Claimed token, claim set to true-> can't be transferred
3. Unclaimed token, claim set to false -> can be transferred
4. Unclaimed token, claim set to true -> can be transferred, token claim boolean set to true and cannot be further transferred

**Pull request merged**